Userland/Applications/VideoPlayer/main.cpp

```diff
@@ -25,32 +25,84 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+                auto result = vp9_decoder.decode_frame(frame);
+                vp9_decoder.dump_frame_info();
+
+                auto const& output_y = vp9_decoder.get_output_buffer_for_plane(0);
+                auto const& output_u = vp9_decoder.get_output_buffer_for_plane(1);
+                auto const& output_v = vp9_decoder.get_output_buffer_for_plane(2);
+                auto y_size = vp9_decoder.get_y_plane_size();
+                auto uv_subsampling_y = vp9_decoder.get_uv_subsampling_y();
+                auto uv_subsampling_x = vp9_decoder.get_uv_subsampling_x();
+                Gfx::IntSize uv_size { y_size.width() >> uv_subsampling_x, y_size.height() >> uv_subsampling_y };
+
+                for (auto y_row = 0u; y_row < video_track.pixel_height; y_row++) {
+                    auto uv_row = y_row >> uv_subsampling_y;
+
+                    for (auto y_column = 0u; y_column < video_track.pixel_width; y_column++) {
+                        auto uv_column = y_column >> uv_subsampling_x;
+
+                        auto y = output_y[y_row * y_size.width() + y_column];
+                        auto cb = output_u[uv_row * uv_size.width() + uv_column];
+                        auto cr = output_v[uv_row * uv_size.width() + uv_column];
+                        // Convert from Rec.709 YCbCr to RGB.
+                        // FIXME: Make some more general color space handling instead of hardcoding this
+                        auto r_float = floorf(clamp(y + (cr - 128) * 219.0f / 224.0f * 1.5748f, 0, 255));
+                        auto g_float = floorf(clamp(y + (cb - 128) * 219.0f / 224.0f * -0.0722f * 1.8556f / 0.7152f + (cr - 128) * 219.0f / 224.0f * -0.2126f * 1.5748f / 0.7152f, 0, 255));
+                        auto b_float = floorf(clamp(y + (cb - 128) * 219.0f / 224.0f * 1.8556f, 0, 255));
+                        auto r = static_cast<u8>(r_float);
+                        auto g = static_cast<u8>(g_float);
+                        auto b = static_cast<u8>(b_float);
+
+                        image->set_pixel(y_column, y_row, Gfx::Color(r, g, b));
+                    }
+                }
+
+                if (result.is_error()) {
+                    outln("Error: {}", result.error().string_literal());
+                    return 1;
+                }
+
+                // FIXME: We can't decode inter frames, and intra frame decoding is incorrect.
+                //        Break out of all loops so that we can display the first frame.
+                //        This loop should also be moved to something that stores demuxer state
+                //        and gets called from the GUI event loop.
+                if (frame_number == 0) {
+                    decoded_frame = true;
+                    break;
+                }
+
+                frame_number++;
             }
+
+            if (decoded_frame)
+                break;
         }
+
+        if (decoded_frame)
```

Comment:
this heavily nested loop construct with the `if (decoded_frame) break;;` chain at the end feels like it would be much nicer to read if it was moved into a helper function where you could early instead of cascading the success result out of each loop.

Reply:
I'll just copy over my later implementation of this since it's cleaner, no point in refactoring this since it was only hacked together to get a frame to display.